File: python/esnn.py
'''
Created on 13/02/2012

@author: stefan
'''

import numpy
import pylab

class SpikeEncoding(object):
    '''
    Abstract class that encapsulates the transformation of real-valued input data
    into trains of spikes
    '''
    
    def __init__(self, **kwargs):
        '''
        '''
        # we store all parameters as class members
        for key, value in kwargs.items():
            setattr(self, key, value)
    
    def convert(self, vec):
        '''
        Converts a real-valued vector into spike times of a group of neurons.
        '''
    
class PopulationEncoding(SpikeEncoding):
    '''
    Implementation of the population encoding using Gaussian receptive fields as
    described in BOTHE et al.
    '''
    
    REQUIRED_PARAMS = ['nb_fields', # number of receptive fields 
                       'beta',      # width of each receptive field 
                       'i_min',     # center of first receptive field   
                       'i_max']     # center of last receptive field
    
    def __init__(self, **kwargs):
        SpikeEncoding.__init__(self, **kwargs)
        
        # check if all parameters have been set by the constructor
        for p in self.REQUIRED_PARAMS:
            if getattr(self, p)==None:
                print "Attribute %s must be set for the PopulationEncoding!" % (p)

        # compute mu and sigma for the Gaussian receptive fields
        fields = numpy.arange(0, self.nb_fields)+1
        self.mu = self.i_min + (2.*fields-3.)/2. * (self.i_max-self.i_min)/(self.nb_fields-2.)
        self.sigma = 1./self.beta * (self.i_max-self.i_min)/(self.nb_fields-2.)

        # create a function for all receptive fields        
        self.rcf = lambda x:  1./(self.sigma * numpy.sqrt(2.*numpy.pi)) * numpy.exp(- (x-self.mu)**2 / (2.* self.sigma**2));
        
        # compute the maximum output of a receptive field for normalization purposes
        self.rcf_max = self.rcf(self.mu[0])[0]
        
    
    def convert(self, vec):
        
        # loop through all values of the vector
        # and compute the response of the receptive fields
        l_response_vec = []
        for v in vec:
            response_vec = self.rcf(v) / self.rcf_max
            l_response_vec.extend(response_vec)
        
        # transform the response into delay times
        a_response_vec = -numpy.array(l_response_vec) + 1.
        
        return a_response_vec
        
        

class ESNN(object):
    '''
    Implementation of the Evolving Spiking Neural Network (eSNN) classification
    algorithm as introduced in REF.
    '''
    
    class TrainedNeuron():
        '''
        Encapsulates a single trained neuron
        '''
        def __init__(self, w, theta, nb_merges, label):
            self.w = w
            self.theta = theta
            self.nb_merges = nb_merges
            
            self.psp = 0.
            self.label = label
            

    def __init__(self, encoder=None, m=0.9, s=0.1, c=0.7):
        '''
        '''
        
        self.encoder = encoder
        self.m = m  # modulation factor
        self.s = s  # similarity threshold
        self.c = c  # ratio of firing threshold
        
        self.repos = {}

        # latest firing time of an input neuron
        self.max_response_time = 1. - 0.1
        
    
    def train(self, data):
        '''
        Trains the eSNN classifier. Note that the training accumulates! Call 
        reset() to start a new training session with an empty neuron repository.
        
        @param data: Training data. Each row represents one sample, each column
            one feature. The last column represents the class label of the sample.
        '''
        
        # last column is the class label
        label = data[:,-1]
        samples = data[:,:-1]
        
        # check if we have individual repositories for all labels already
        for l in set(label):
            if l not in self.repos.keys():
                # no, we don't know this label, yet, so we add it
                self.repos.update({l:[]})
        
        for i,sample in enumerate(samples):
            
            # load label
            l = label[i]
            
            # convert sample into spike times
            spikes = self.encoder.convert(sample)
            
            # compute the weights, max PSP and the firing threshold
            w, theta = self._train_sample(spikes)
            
            # find similar neurons
            neuron = self._find_similar(w, l)
        
            # if we already have trained a similar neuron earlier,
            # then merge with the one we have just trained
            if neuron:
                self._merge(neuron, self.TrainedNeuron(w, theta, 0, None))
            # otherwise, we just add the newly trained neuron into the repository    
            else:
                self.repos[l].append(self.TrainedNeuron(w, theta, 1, l))

    def _merge(self, neuron1, neuron2):
        '''
        Merges the weight vectors and the firing threshold of the two neurons
        and writes the result back to neuron1.
        '''
        neuron1.w = (neuron2.w + neuron1.nb_merges*neuron1.w) / (1.+neuron1.nb_merges)
        neuron1.theta = (neuron2.theta + neuron1.nb_merges*neuron1.theta) / (1.+neuron1.nb_merges)
        neuron1.nb_merges += 1

    def _find_similar(self, weights, label):
        '''
        Searches for a similar neuron in the repository of trained neurons
        that belong to the specified class. 
        '''
        
        # first we check, if there is any neuron in the repository belonging
        # to the given class
        if len(self.repos[label])==0:
            # if there is no neuron in the repository, we can stop here
            return None
        
        # find the neuron with the smallest distance to the given weight vector 
        l_dist = []
        for neuron in self.repos[label]:
            l_dist.append(numpy.linalg.norm(neuron.w - weights))
        min_index = numpy.argmin(l_dist)
        min_dist = l_dist[min_index]
        
        if min_dist<self.s:
            return self.repos[label][min_index]
        else:
            return None
        
    
    def _train_sample(self, sample):
        '''
        Computes the weights of a training neuron based on the given input
        sample.
        
        Returns a tuple (weights, theta) representing the trained weight
        vector along with the firing threshold theta of the neuron.
        '''
        index = numpy.argsort(sample)
        weights = numpy.zeros(len(sample))
        weights[index] = self.m**numpy.arange(len(sample))

        # correct the weights for too late spikes
        index = pylab.find(sample > self.max_response_time)
        weights[index] = 0.
        
        u_max = numpy.sum(weights**2)
        theta = self.c * u_max
        
        return weights, theta
    
    def reset(self):
        '''
        Resets the repository of trained neurons.
        '''
        # remove all neurons from the repository and the neuron list
        self.all_neurons = []
        for k in self.repos.keys():
            del self.repos[k]
        del self.repos
        
    
    def test(self, data):
        '''
        Tests the eSNN classifier.
        
        @param data: Testing data. Each row represents one sample, each column
            one feature. The last column represents the class label of the sample.
        '''
        
        # last column is the class label
        label = data[:,-1]
        samples = data[:,:-1]
        
        # check if we have individual repositories for all labels already
        for l in set(label):
            if l not in self.repos.keys():
                # no, we don't know this label, yet, so we add it
                self.repos.update({l:[]})

        # update the list of all neurons
        self.all_neurons = []
        for k in self.repos.keys():
            self.all_neurons.extend(self.repos[k])

        # create a matrix of all weight vectors, all firing thresholds and all PSPs
        # NOTE: this step intends to accelerate the propagation method _propagate_fast()
        self.weight_matrix = []
        self.all_theta = [] 
        for n in self.all_neurons:
            self.weight_matrix.append(n.w)
            self.all_theta.append(n.theta)
        self.weight_matrix = numpy.array(self.weight_matrix)
        self.all_theta = numpy.array(self.all_theta)
        self.all_psp = numpy.zeros(len(self.all_neurons))
        
        # prepare results
        nb_correct = 0
        classified_labels = []
        
        # loop over all samples of the data set        
        for i,sample in enumerate(samples):
            
            # load label
            l = label[i]
            
            # convert sample into spike times
            spikes = self.encoder.convert(sample)

            # find neuron that spikes first
            result = self._propagate_fast(spikes)
            if result:
                # the sample was classified into some known class
                # extract results and store
                neuron, spike_time, class_label = result
                
                # store classification result
                classified_labels.append(class_label)
                    
                # check the classification result
                if class_label == l:
                    nb_correct += 1
            else:
                # the sample was not classified (none of the neurons emitted a spike)
                # store classification result
                classified_labels.append(-1)
            
        # return the list of classifications and the classification accuracy
        return classified_labels, float(nb_correct)/float(len(data))
    
    def _propagate_fast(self, sample):
        '''
        Propagates the given input through the network and determines which
        neuron spikes first.
        
        Returns a tuple containing the first-spiking neuron, its spike time 
        and the corresponding class label. 
        '''

        # reset all PSPs        
        self.all_psp[:] = 0. 
                
        # event driven processing of the input sample
        # we loop over all sorted firing times of the input
        sorted = numpy.argsort(sample)
        for order, index in enumerate(sorted):
            #update the PSP of all neurons in the repository
            self.all_psp += self.weight_matrix[:,index] * self.m**order 
            
            # check threshold condition
            active_neurons = pylab.find(self.all_psp > self.all_theta)
            if len(active_neurons) > 0:
                # stop the simulation if any neuron has fired
                return self.all_neurons[active_neurons[0]], sample[index], self.all_neurons[active_neurons[0]].label
        
        # no neuron has fired
        return None

    
    def create(self, repos):
        '''
        Creates a trained repository of neurons based on the specified repository.
        '''
        self.reset()          # clean up
        self.repos = repos    # store the new repository


    def _propagate(self, sample):
        '''
        @deprecated: Use the much faster _propagate_fast() method instead
        
        Propagates the given input through the network and determines which
        neuron spikes first.
        
        Returns a tuple containing the first-spiking neuron, its spike time 
        and the corresponding class label. 
        '''
        
        # reset the psp of all neurons
        for n in self.all_neurons:
            n.psp = 0.
            
        # event driven processing of the input sample
        # we loop over all sorted firing times of the input
        sorted = numpy.argsort(sample)
        for order, index in enumerate(sorted):
            #update the state of all neurons in the repository
            for n in self.all_neurons:
                # accumulate the PSP
                n.psp += n.w[index]*self.m**order
                
                # check threshold condition
                if n.psp > n.theta:
                    # stop the simulation once a neuron has fired
                    return n, sample[index], n.label
        return None

File: ESNN.R
##
#   Class:  ESNN
#
# Description:
#   Implementation of the Evolving Spiking Neural Network (eSNN) classification
#   algorithm as introduced in REF.
#
# @author: Vasilis Bankov
##

ESNN <- R6Class("ESNN",
  public = list(
    encoder = NA,
    m = NA, # modulation factor
    s = NA, # similarity threshold
    c = NA, # ratio of firing threshold
    repos = list(),
    maxResponseTime = .9, # latest firing time of an input neuron

    # Initialization
    initialize = function(encoder, m, s, c){
      stuffIsCool <- private$checkInitializationConsistency(encoder, m, s, c)
      if(!stuffIsCool){
        # Something's missing
        cat('---\tOops !!\t---\n-\tThere was an error initializing SpikeEncoding. An argument seems to be missing.')
      }else{
        self$encoder <<- encoder
        self$m <<- m
        self$s <<- s
        self$c <<- c
      }
    },

    #   Trains the eSNN classifier. Note that the training accumulates! Call
    #   reset() to start a new training session with an empty neuron repository.

    #   @param data: Training data. Each row represents one sample, each column
    #       one feature. The last column represents the class label of the sample.
    train = function(data){
      labls <- as.data.frame(data[,length(data)])
      sampls <- as.data.frame(data[,1:length(data)-1])


      # we'll save the unique labels as string in our neuron repositories
      u <- unique(labls)[[1]]
      for(i in 1:length(u)){
        n <- Repo$new(as.character(u[i]))
        private$repoLabels <- c(private$repoLabels, as.character(u[i]))
        self$repos <- c(self$repos, n)
      }


      for(i in 1:length(sampls[,1])){
        sampl <- sampls[i,]

        # load label
        l <- labls[i,]

        # convert sample into spike times
        spikes <- self$encoder$convert(sampl)

        # compute the weights, max PSP and the firing threshold
        params <- private$trainSample(spikes)
        w <- params$w
        theta <- params$theta

        # find similar neurons
        similarNeuron <- private$findSimilar(w, l)

        if(!is.null(similarNeuron)){
          newNeuron <- TrainedNeuron$new(w=w, theta = theta, 0, NULL)
          mergedNeuron <- private$merge(similarNeuron$neuron, newNeuron)
          ll <- which(private$repoLabels==mergedNeuron$label)
          self$repos[[ll]]$neurons[[similarNeuron$index]] <- NULL
          self$repos[[ll]]$neurons[[similarNeuron$index]] <- mergedNeuron
        }else{
          ll <- which(private$repoLabels==l)
          self$repos[[ll]]$append(TrainedNeuron$new(w, theta, 1, l))
        }

      }

    },

    test = function(data){
      # Tests the eSNN classifier.

      # @param data: Testing data. Each row represents one sample, each column
      #     one feature. The last column represents the class label of the sample.

      # last column is the class label
      label <- as.data.frame(data[,ncol(data)])
      samples <- as.data.frame(data[,1:ncol(data)-1])

      # check if we have individual repositories for all labels already
      for (l in label) {
        if(length(which(private$repoLabels==l))<0){
          # no, we don't know this label, yet, so we add it
          private$repoLabels <- c(self$repoLabels, l)
        }
      }

      # update the list of all neurons
      for(k in private$repoLabels){
        kk <- which(private$repoLabels==k)
        private$all_neurons <- c(private$all_neurons, self$repos[[kk]]$neurons)
      }


      # create a matrix of all weight vectors, all firing thresholds and all PSPs
      # NOTE: this step intends to accelerate the propagation method propagate_fast()
      private$weight_matrix <- matrix(NA, nrow=length(private$all_neurons), ncol=length(private$all_neurons[[1]]$w))
      counter <- 1
      for(n in private$all_neurons){
        private$weight_matrix[counter,] <- n$w
        private$all_theta <- c(private$all_theta, n$theta)
        counter <- counter + 1
      }
      private$all_psp <- matrix(0,length(private$all_neurons))

      # prepare results
      nb_correct <- 0
      classified_labels <- c()
      resultsTable <- data.frame()

      # loop over all samples of the data set
      for(i in 1:length(samples[,1])){

        # load label
        testLabel <- label[i,]
        smpl <- samples[i,]

        # convert sample into spike times
        spikes <- self$encoder$convert(smpl)

        # find neuron that spikes first
        result <- private$propagate_fast(spikes)
        if(!is.null(result)){
          ## the sample was classified into some known class
          ## extract results and store

          neuron <- result$neuron
          spike_time <- result$spike_time
          class_label <- result$class_label

          ## store classification result
          classified_labels <- c(classified_labels, class_label)
          resultsTable <- rbind(resultsTable, data.frame(testLabel, class_label, class_label == testLabel) )
          ## check the classification result
          #cat(paste('\nLabel: ',class_label, ' correct: ',testLabel, ' ',(class_label == testLabel) ))
          if(class_label == testLabel){
            nb_correct <- nb_correct+1
          }

        }else{
          ## the sample was not classified (none of the neurons emitted a spike)
          ## store classification result
          classified_labels <- c(classified_labels, -1)
        }
      }

      ## return the list of classifications and the classification accuracy
      #return(list("classified_labels"=classified_labels, "accuracy"=nb_correct/length(data[,1])))
      names(resultsTable) <- c('label','classified','result')
      conf <- confusionMatrix(resultsTable$classified, resultsTable$label)

      confTable <- conf$table
      precision <- NaN
      if(dim(confTable)[[1]]<3){  # Only two classes are present, we'll calculate the precision
        b<- confTable[1,2]
        d<- confTable[2,2]
        cat('2 classes found, appending precision.\n')
        precision <- d / (b+d)
      }
      accuracy <- conf$overall[[1]]
      return(list(results=resultsTable, stats=conf, precision=precision, accuracy=accuracy))
    }

  ),

  private = list(
    weight_matrix = matrix(),
    all_neurons = c(),
    all_theta = c(),
    all_psp = NA,

    propagate_fast = function(sampl){
      # Propagates the given input through the network and determines which
      # neuron spikes first.

      # Returns a list containing the first-spiking neuron, its spike time
      # and the corresponding class label.

      # reset all PSPs
      private$all_psp <- matrix(0,length(private$all_neurons))

      # event driven processing of the input sample
      # we loop over all sorted firing times of the input
      sorted <- order(sampl)
      ordr <- 0
      for(index in sorted){
        #ordr <- sampl[index]
        #update the PSP of all neurons in the repository
        private$all_psp <- private$all_psp + private$weight_matrix[,index] * (self$m^ordr)

        # check threshold condition
        active_neurons <- which(private$all_psp > private$all_theta)

        if(length(active_neurons)>0){
          first_active_neuron_index <- active_neurons[1]
          first_active_neuron <- private$all_neurons[[first_active_neuron_index]]
          #cat(paste('fired neurons: ',length(active_neurons),' \tfirst_active_neuron_index:',first_active_neuron_index,'\n'))

          return(list("neuron"=first_active_neuron, "spike_time"=sampl[index], "class_label"= first_active_neuron$label ))
        }
        ordr <- ordr+1
      }
      #cat(paste('no neurons fired \n'))
      return(NULL)
    },

    merge = function(neuron1, neuron2){
      neuron1$w <- ( neuron2$w + neuron1$nbMerges*neuron1$w ) / ( 1 + neuron1$nbMerges )
      neuron1$theta <- ( neuron2$theta + neuron1$nbMerges * neuron1$theta ) / ( 1 + neuron1$nbMerges )
      neuron1$nbMerges <- neuron1$nbMerges+1
      return(neuron1)
    },

    repoLabels = c(),

    checkInitializationConsistency = function(encoder, m, s, c){ # check initialization consistency
      return( !missing(encoder) && !missing(m) && !missing(s) && !missing(c)  ) # no argument should be missing
    },

    trainSample = function(sample){
      # Computes the weights of a training neuron based on the given input sample.

      #   Returns a tuple (weights, theta) representing the trained weight
      #   vector along with the firing threshold theta of the neuron.
      index <- order(sample)
      weights <- matrix(0, length(sample)) # r equivalent to numpy.zeros(x)

      weights[index] <- self$m ^ (seq(length(sample))-1)

      # correct the weights for too late spikes
      weights[ sample[ sample > self$maxResponseTime ] ] <- 0

      u_max <- sum(weights^2)
      theta <- self$c * u_max

      return(list("weights"=weights, "theta"=theta))
    },

    findSimilar = function(weights, label) {
      # Searches for a similar neuron in the repository of trained neurons
      # that belong to the specified class.

      # first we check, if there is any neuron in the repository belonging
      # to the given class
      ll <- which(private$repoLabels==label)
      if(length(self$repos[[ll]]$neurons)==0){
        return(NULL)
      }

      l_dist <- c()
      o<-1
      for(neuron in self$repos[[ll]]$neurons){
        # linear norm
        nnn <- norm(as.matrix(c(neuron$w - weights)), type="F")
        l_dist <- c(l_dist, nnn)
      }
      min_index <- which.min(l_dist)
      min_dist <- min(l_dist)

      if(min_dist < self$s){
        return(c(label= label, index= min_index, neuron=self$repos[[ll]]$neurons[min_index]))
      }else{
        return(NULL)
      }

    }

  )
)